crates/flox-rust-sdk/src/models/manifest.rs

```diff
@@ -0,0 +1,202 @@
+) -> Result<PackageInsertion, TomlEditError> {
+    debug!("attempting to insert packages into manifest");
+    let mut already_installed: HashMap<String, bool> = HashMap::new();
+    let mut toml = manifest_contents
+        .parse::<Document>()
+        .map_err(TomlEditError::ParseManifest)?;
+    match toml.entry("install") {
+        toml_edit::Entry::Occupied(ref mut existing_installs) => {
+            debug!("editing existing [install] table");
+            if let Item::Table(ref mut installs) = existing_installs.get_mut() {
+                for pkg in pkgs {
+                    if !installs.contains_key(&pkg) {
+                        installs.insert(&pkg, Item::Value(Value::InlineTable(InlineTable::new())));
+                        already_installed.insert(pkg.clone(), false);
+                        debug!("package {pkg} newly installed");
+                    } else {
+                        already_installed.insert(pkg.clone(), true);
+                        debug!("package {pkg} already installed");
+                    }
+                }
+
+                // TODO: Figure out a better sorting system
+                // installs.sort_values_by(|key1, _, key2, _| key1.cmp(key2));
+            } else {
+                return Err(TomlEditError::MalformedInstallTable(
+                    existing_installs.get().type_name().into(),
+                ));
+            }
+        },
+        toml_edit::Entry::Vacant(empty_installs) => {
+            debug!("creating new [install] table");
+            let mut installs_table = Table::new();
+            for pkg in pkgs {
+                installs_table.insert(&pkg, Item::Value(Value::InlineTable(InlineTable::new())));
+                already_installed.insert(pkg.clone(), false);
+            }
+            // TODO: Figure out a better sorting system
+            // installs_table.sort_values_by(|key1, _, key2, _| key1.cmp(key2));
+            empty_installs.insert(Item::Table(installs_table));
+        },
+    };
+    Ok(PackageInsertion {
+        new_toml: if !already_installed.values().all(|p| *p) {
+            Some(toml)
+        } else {
+            None
+        },
+        already_installed,
+    })
+}
+
+// FIXME: will be used in uninstall
+/// Check whether a TOML document contains a line declaring that the provided package
+/// should be installed.
+#[allow(unused)]
+pub fn contains_package(toml: &Document, pkg_name: &str) -> Result<bool, TomlEditError> {
+    if let Some(installs) = toml.get("install") {
+        if let Item::Table(installs_table) = installs {
+            Ok(installs_table.contains_key(pkg_name))
+        } else {
+            Err(TomlEditError::MalformedInstallTable(
+                installs.type_name().into(),
+            ))
+        }
+    } else {
+        Ok(false)
+    }
+}
+
+/// List the packages contained in the contents of a manifest.
+pub fn list_packages(manifest_contents: &str) -> Result<Option<Vec<String>>, TomlEditError> {
```

Comment:
question: why return an `Option<Vec<_>>` over just `Vec<_>` if "absense of an `install` key is rather irrelevant for us?

Reply:
I don't understand what you mean by
> if "absense of an install key is rather irrelevant for us?

Do you mean that if there's no `install` table you would prefer to send an empty `Vec`? If there are no packages I'd rather not create an empty `Vec` for no reason.

Reply:
Yeah I would expect an empty Vec if there is no package installed..

but feel free to armadillo on this, i don't think it's worth to bikeshed over